Only load depsfile if not dirty [Fix #2666]

moritzx22 opened the pull request:
This Merge request is related to [#2666](https://github.com/ninja-build/ninja/issues/2666).

This pull request proposes to load the depsfile only if it is not dirty.
For a more detailed description, see [comment #2666](https://github.com/ninja-build/ninja/issues/2666#issuecomment-3394810561)

mathstuf reviewed (changes requested):
I think the test suite leak fix can be its own PR. Can a test case for the compelling scenario be added?

moritzx22 commented on src/build_test.cc:
```diff
@@ -534,6 +534,17 @@ struct FakeCommandRunner : public CommandRunner {
   VirtualFileSystem* fs_;
 };
 
+class SafeRelase {
```
done

moritzx22 commented:
> think the test suite leak fix can be its own PR

new PR created: #2684. I will keep this PR unchanged, until PR2684 is merged and a rebase can be done.

mathstuf reviewed (commented):
Looks OK to me, but someone else should also review (I did it mainly from the test cases).

moritzx22 commented on src/missing_deps.cc:
```diff
@@ -39,16 +39,17 @@ struct NodeStoringImplicitDepLoader : public ImplicitDepLoader {
-                                  std::string* err);
+  bool ProcessDepfileDeps(Edge* edge, std::vector<StringPiece>* depfile_ins,
+                          std::string* err,
+                          std::array<std::size_t, 2>* input_range) override;
```
Marking this as `override` is a candidate for separate PR.
Without this commit, the unit-tests do pass. Test-coverage could be improved to make the unittests fail if the function is not overriden.

mathstuf replied:
Make the base class pure virtual (it can keep the implementation)?

moritzx22 replied:
> Make the base class pure virtual (it can keep the implementation)?

Class `ImplicitDepLoader` would become an abstract base class, but it is still required to instantiate objects of this class. That would not work anymore.

moritzx22 committed:
Load depfile only when output nodes are not dirty

This change updates the logic so that the depfile is loaded only when no
output node is not dirty based on manifest and dyndep inputs (excluding the
depfile itself). If the outputs are already scheduled to be regenerated, loading
the depfile is unnecessary, since its only purpose is to trigger regeneration
of the outputs — which is already guaranteed.

Avoiding the use of a potentially outdated depfile prevents incorrect cycle
detection and ensures that stale depfiles are no longer added to the graph.

moritzx22 commented:
Rebased to master in previous push.

digit-google reviewed (changes requested):
This PR is impossible to review properly as a stack of 8 commits with what looks like random changes with unclear commit messages, please squash / rebase this into something that is simpler to review. For example:

- one commit to add the constness changes + the cp-deps test rule implementation.

- second commit to change the LoadDepXXX() signatures with proper documentation of all new parameters, preferably without changing the implementation yet.

- a third commit that changes the implementation to change the behavior / fix the bug and modify the tests accordingly.

Each commit should have a clear commit message explaining its purpose and why things are changed in a certain way. I'll add some inline comments too.

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+namespace {
+
+/// execute hash only once in lifetime of object and only on request
+struct hashCommand {
```
Please follow existing coding conventions, i.e. struct/class names should use PascalCase (hashCommand -> HashCommand), and member variables should use trailing underscore (valid -> valid_). Moreover, call this LazyEdgeCommandHash for clarity.

moritzx22 replied:
done

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+
+/// class is similar to a pointer of BuildLog::LogEntry
+/// additionally the LookupByOutput is cached for performance reasons
+class LogEntryCache {
```
nit: Name this CachedLogEntry for clarity since this is not a cache.

moritzx22 replied:
done

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+ public:
+  LogEntryCache(){};
+
+  operator bool() const { return entry_; }
```
Explain what this corresponds to and when it is safe to call, since it never looks at evaluated_, the meaning of the result value is ambiguous. Consider replacing this with `is_valid()` for clarity, bool operators can lead to surprising bugs.

moritzx22 replied:
renamed to `is_valid()`

moritzx22 replied:
example usage for clarity
```cpp
CachedLogEntry cached;
if(cached.is_valid()) cached->foo(); // nullptr
cached.LookupByOutput(build_log, output);  // assign value
if(cached.is_valid()) cached->foo(); // call foo()
cached.LookupByOutput(build_log, output);  // already cached
if(cached.is_valid()) cached->foo(); // call foo()

// a raw pointer instead
BuildLog::LogEntry* entry = nullptr;
if(entry) entry->foo(); // nullptr
entry = build_log.LookupByOutput(output);
if(entry.is_valid()) entry->foo(); // call foo()
entry = build_log.LookupByOutput(output); // second call
if(entry.is_valid()) entry->foo(); // call foo()

```

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+  BuildLog::LogEntry* entry_ = nullptr;
+};
+
+bool LogEntryCache::LookupByOutput(const BuildLog* buildLog, const Node* output) {
```
This interface is ambiguous, because the function could in theory be called with different |output| values and will only return a result corresponding to the first call. You could implement something similar without a dedicated LogEntryCache class with a simple std::map<const Node*, const BuildLog::LogEntry*> instead inside RecomputeOutputsDirty_, which would be simpler / clearer.

moritzx22 replied:
Here are the reasons behind introducing this new class:

Requirements
 - Cache the result of LookupByOutput.
 - Store only a single pointer and a single boolean per cache entry.
 - Keep the cache in contiguous memory (std::vector) for compiler‑friendly access patterns.
 - Allocate the memory only once (vector sized in the constructor).
 - Allow lookup of the cached value for a given output at effectively zero cost.

`RecomputeOutputsDirty` is performance‑critical.
Its worst‑case scenario is a clean build, where no early exits occur and every output must be visited. `RecomputeOutputsDirty` will be invoked twice in this situation because the depfile is loaded here. This is exactly where the cache provides the most benefit.

To achieve this performance, the class relies on strict usage assumptions:
`LookupByOutput` must always be called with the same parameters for a given instance. Debug assertions enforce this, and the assumptions are documented in the code. These constraints allow the implementation to remain efficient.

`RecomputeOutputsDirtyCache` ensures these assumptions hold. It selects the correct output and its associated cache entry for processing. The `CachedLogEntry` type is defined in the private section to prevent accidental misuse outside the intended context.

A `std::map` could also be used to implement the cache, and it would likely be simpler to write, but I expect its performance to be worse, especially for clean builds.

Please advise.

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+}
+
+/// performance optimized to recompute the outputs
+class RecomputeOutputsDirty_ {
```
Please do not use trailing underscores in class names (or even inside them). Call this RecomputeOutputsDirtyCache instead, or something similar. Also consider moving changes related to performance optimizations to their own commit so they can be reviewed more easily.

moritzx22 replied:
done

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+/// performance optimized to recompute the outputs
+class RecomputeOutputsDirty_ {
+ public:
+  RecomputeOutputsDirty_(BuildLog* buildLog, OptionalExplanations& explanations,
```
coding style: please use snake_case for variable / member identifiers (buildLog -> build_log)

moritzx22 replied:
done

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+                         Edge* edge)
+      : buildLog_(buildLog), explanations_(explanations), edge_(edge),
+        LogEntry_(edge->outputs_.size()) {}
+  bool all(const Node* most_recent_input);
```
nit: document what these methods do.

moritzx22 replied:
done

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+  return false;
+}
+
+// disable warning for windows
```
Explain why these are needed exactly.

moritzx22 replied:
MSVC warns (and errors) on constructs like:
```
if (false) {
// do some stuff
}
```
This is suppressed. Anyhow this is obsolete with the change to c++17 and the use of `if constexpr`

digit-google replied:
This should go as a comment inside the source code, so that future maintainers now how / when to keep this.

moritzx22 replied:
sorry for not being clear. In the latest push it looks like

```cpp
if constexpr (false) {
// do some stuff
}
```
and no warning or error is reported by MSVC anymore. The disable warning stuff has been removed.

digit-google commented on src/graph.cc:
```diff
@@ -31,6 +31,217 @@
+  assert(FIRSTRUN || !(cond)); /* NOLINT */ \
+  if (FIRSTRUN && (cond))      /* NOLINT */
+
+template <bool FIRSTRUN>
```
I strongly recommend to getting rid of the template parameter, and adding a simple `first_run` function parameter instead.

moritzx22 replied:
The template parameter is `constexpr`, which gives the compiler the best opportunity to optimize the code. Conceptually, the template function represents two distinct functions and helps avoid code duplication. The runtime `if` is replaced with `if constexpr` in the next push, so the unused branch is removed entirely at compile time.

There are essentially three design options:
 - A template function  
   - Uses a `constexpr template parameter` to generate two optimized code paths without duplication.
   - Cleanly separates the regular function parameters from the compile‑time selection parameter.
 - A regular function with a runtime parameter  
   - Simpler interface
 - Two separate functions  
    - Maximum clarity, but duplicates code.

Please restate your preference.
If you still recommend avoiding the template parameter, I’d appreciate some more detail on why the template approach is undesirable in this context.

Note: The `if` is changed to `if constexpr` in the macro.

digit-google commented on src/graph.h:
```diff
@@ -295,7 +296,9 @@ struct ImplicitDepLoader {
   //                          or out of date).
-  bool LoadDeps(Edge* edge, std::string* err);
+  bool LoadDeps(Edge* edge, std::string* err,
+                std::array<std::size_t, 2>* input_range = nullptr);
```
Please update the documentation to explain the purpose of this new `input_range` parameter. Consider using a simple `struct InputRange { size_t start; size_t end; }` definition to make this easier to read and understand. Clarify that this is an optional output parameter.

moritzx22 replied:
replaced `std::array` with:
```
struct InputView {
  std::size_t offset_begin = 0;
  std::size_t offset_end = 0;
};
```

moritzx22 replied:
The original `conditional` was removed to improve performance.
The Call that previously used that `conditional` now uses a dummy object.
This path(missingdeps) isn't performance‑critical, so using a dummy is acceptable.

digit-google commented on src/graph.cc:
```diff
@@ -79,6 +290,46 @@ bool DependencyScan::RecomputeDirty(Node* initial_node,
+      return false;
+  }
+
+  const auto input_end = edge->inputs_.cend() - input_range[1];
```
Is this computation correct here? Can you clarify the meaning of input_range[1]? From the name "range" it can be assumed that this would be the position of the first item after the range, but in this case, you would use `input_end = edge->inputs_.cbegin() + input_range[1]` instead.

If the value is a count instead, "input_span" might be a better name, but the computation would be `input_end = edge->inputs_.cbegin() + input_range[0] + input_range[1]` so I am puzzled as to what this code does.

digit-google commented on src/graph.cc:
```diff
@@ -146,70 +395,70 @@ bool DependencyScan::RecomputeNodeDirty(Node* node, std::vector<Node*>* stack,
   // Load output mtimes so we can compare them to the most recent input below.
-  for (Node* o : edge->outputs_) {
+  for (vector<Node*>::iterator o = edge->outputs_.begin();
+       o != edge->outputs_.end(); ++o) {
```
nit: Why regress here when the original code was perfectly fine?

moritzx22 replied:
corrected

digit-google commented on src/graph.cc:
```diff
@@ -146,70 +395,70 @@ bool DependencyScan::RecomputeNodeDirty(Node* node, std::vector<Node*>* stack,
+  // if an rebuild is necessary the deps log is outdated for this target
+  if (!edge->deps_loaded_ && !dirty) {
+    // This is our first encounter with this edge.  Load discovered deps.
+    std::array<std::size_t, 2> newLinks{ 0, 0 };
```
nit: `newLinks` doesn't mean anything here. Use something more specific here. `new_deps` maybe?

digit-google replied:
Or more precisely `new_deps_range`

moritzx22 replied:
renamed to `new_deps`

digit-google commented on src/graph.cc:
```diff
@@ -734,10 +912,14 @@ bool ImplicitDepLoader::ProcessDepfileDeps(
   }
 
+  if (input_range)
+    (*input_range)[1] = std::distance(implicit_dep, edge->inputs_.end());
```
oh, so you are storing the distance from the last implicit input to the end of the array. This data structure definitely is not a range. Consider storing the distance from the start for clarity.

moritzx22 replied:
This data structure represents a subset of a container.
The usual approach starting from index 0 has a drawback:
the default‑constructed view should represent the entire container, which requires knowing its size.

In this context it would look  like:
```cpp
if (!RecomputeEdgesInputsDirty(node, InputView(), most_recent_input, dirty,
                                 stack, validation_nodes, err))
// would need to become
if (!RecomputeEdgesInputsDirty(node, InputView{0, node->in_edge()->inputs_.size()}, most_recent_input, dirty,
                                 stack, validation_nodes, err))
```
[graph.cc#L482](https://github.com/moritzx22/ninja/blob/0070603c01cb1febf33a24108e80d128882e39e2/src/graph.cc#L482)

I can implement the change, and it can certainly be expressed more cleanly than in the example above. However, in this particular context the change makes the code a bit more complicated. Before proceeding, I’d appreciate your advice.

digit-google replied:
Second, there is 

> This data structure represents a subset of a container. The usual approach starting from index 0 has a drawback: the default‑constructed view should represent the entire container, which requires knowing its size.
> 
Technically, this is neither a "view" nor a "subset" as these terms usually refer to objects that can be used directly to access individual items. This is not the case here: you just have a pair of numbers, whose interpretation requires additional information (in this case the exact and unmodified `inputs_` array they refer to). Hence using something like "range" in the name makes more sense. Another option is to store the  `Edge` pointer in the data structure (or at least a pointer to its edge->inputs_ array).

If you prefer a non-conventional layout / interpretation for the values, I strongly recommend making a custom class with human-friendly accessors to properly document its purpose and simplify its usage. E.g.

```
struct EdgeInputsRange {
  /// Create new instance covering all |edge| inputs.
  EdgeInputsRange(const Edge* edge);
  
  /// Create instance covering the [start_pos..end_pos) interval of |edge| inputs.
  EdgeInputsRange(const Edge* edge, size_t start_pos, size_t end_pos);

  size_t start_pos() const;
  size_t end_pos() const;

private:
  ...
};
```

> In this context it would look like:
> 
> ```c++
> if (!RecomputeEdgesInputsDirty(node, InputView(), most_recent_input, dirty,
>                                  stack, validation_nodes, err))
> // would need to become
> if (!RecomputeEdgesInputsDirty(node, InputView{0, node->in_edge()->inputs_.size()}, most_recent_input, dirty,
>                                  stack, validation_nodes, err))
> ```
> 
> [graph.cc#L482](https://github.com/moritzx22/ninja/blob/0070603c01cb1febf33a24108e80d128882e39e2/src/graph.cc#L482)
> 
> I can implement the change, and it can certainly be expressed more cleanly than in the example above. However, in this particular context the change makes the code a bit more complicated. Before proceeding, I’d appreciate your advice.

moritzx22 replied:
changed to 
```cpp
struct EdgeInputsRange {
  /// Create new instance covering all |edge| inputs.
  EdgeInputsRange(Edge* edge)
      : beg_(edge->inputs_.begin()), end_(edge->inputs_.end()),
        edge_(edge) {}

  using iterator = std::vector<Node*>::const_iterator;

  iterator begin() const { return beg_; }
  iterator end() const { return end_; }

  iterator beg_;
  iterator end_;

  /// The edge whose input range is being viewed.
  Edge* const edge_;
};
```
in recent push. See https://github.com/ninja-build/ninja/pull/2680#issuecomment-4209133391

moritzx22 committed:
Reorder operations in RecomputeNodeDirty

This change adjusts the internal order of the load output mtimes step and the
step that recomputes the dirty state of inputs. The modification does not alter
any functional behavior. The reordering improves internal consistency and
prepares the code for upcoming changes.

moritzx22 committed:
Refactoring

This change refactors internal parts of the code without altering functional
behavior. It prepares the implementation for a future update in which
dependencies will be loaded only when inputs are not marked dirty.

The sequence in which only a subset of inputs can be specified to be processed
will matter for upcoming changes to depfile loading. A new helper function
'RecomputeEdgesInputsDirty' has been introduced for clarity.
The function can be specified to visit only a subset of an edge’s inputs as
well.

moritzx22 committed:
Refactor dirty‑state recomputation with caching and cleanup

Replace DependencyScan::RecomputeOutputDirty with the new
RecomputeOutputsDirtyCache helper class to centralize the
logic for determining whether edge outputs are dirty.
The new cache‑aware implementation avoids redundant work,
improves readability, and prepares the codebase for upcoming
changes to load depfiles only when nodes are not dirty.

Build‑log lookups are now cached, and command hash computation
is performed lazily to improve performance.

This commit introduces no functional changes.

moritzx22 commented:
> This PR is impossible to review properly as a stack of 8 commits with what looks like random changes with unclear commit messages, please squash / rebase this into something that is simpler to review. ...

True. I will reorder and clean up the commits so that each one has a clear purpose and is easier to understand. Only the final commit introduces a functional change. All earlier commits are refactoring or cleanup and keep current master behavior. After restructuring the history, this separation will be clearer and the review much simpler. Most of the other comments will be incorporated as well.

moritzx22 commented:
next push, see https://github.com/ninja-build/ninja/pull/2680#discussion_r3053771337

moritzx22 commented:
### changes next push
- minor improvments
- ImplicitDepLoader::LoadDepFileTry()
  - use Stat instead of Readfile
- ImplicitDepLoader::LoadDepsFromLogTry()
  - call StatIfNecessary()

digit-google commented on src/graph.cc:
```diff
@@ -269,7 +295,8 @@ bool DependencyScan::VerifyDAG(Node* node, vector<Node*>* stack, string* err) {
 
 bool DependencyScan::RecomputeOutputsDirty(Edge* edge, Node* most_recent_input,
-                                           bool* outputs_dirty, string* err) {
+                                           bool* const outputs_dirty,
```
nit: drop the `const` here as it does nothing.

moritzx22 replied:
done

digit-google replied:
I am still seeing it?

moritzx22 replied:
You're right, I overlooked that. I'll correct it. See https://github.com/ninja-build/ninja/pull/2680#issuecomment-4273506744

digit-google commented on src/graph.cc:
```diff
@@ -715,14 +744,16 @@ bool ImplicitDepLoader::LoadDepFile(Edge* edge, const string& path,
   vector<Node*>::iterator implicit_dep =
-      PreallocateSpace(edge, static_cast<int>(depfile_ins->size()));
+      PreallocateSpace(input_range->edge_, static_cast<int>(depfile_ins->size()));
+  input_range->beg_ = implicit_dep;
```
Modifying members of this instance directly breaks encapstulation and is bad style. Use a constructor with explicit arguments for begin/end instead to make this clearer.

The input_range argument should appear after input arguments in the function signature, and the documentation should clarify that it is written to on success.

moritzx22 replied:
Respective change has been reverted, comment should be obsolete. see [comment](https://github.com/ninja-build/ninja/pull/2680#discussion_r3089414978).

digit-google commented on src/graph.h:
```diff
@@ -295,7 +296,7 @@ struct ImplicitDepLoader {
   /// Load implicit dependencies for \a edge.
```
Update the document to clarify the role of |input_range| here. It would be clearer if you keep the Edge* parameter as input, and make|input_range| and output-only parameter after it instead (this avoids dealing with partially-constructed objects which make the code harder to understand).

moritzx22 replied:
```cpp
  /// Load implicit dependencies for \a edge.
  /// @return emtpy optional on error (without filling \a err if info is just
  ///                                  missing or out of date).
  /// @return newly added `EdgeInputsRange` on success, describing the
  ///         dependencies that were added to the graph.
  std::optional<EdgeInputsRange> LoadDeps(Edge* edge, std::string* err);
  //                        bool LoadDeps(Edge* edge, std::string* err);
```
new signature is more clear. Outputs are now separated from inputs.

digit-google commented on src/graph.h:
```diff
@@ -304,17 +305,18 @@ struct ImplicitDepLoader {
  protected:
   /// Process loaded implicit dependencies for \a edge and update the graph
```
Again clarify the rold of |input_range|. In this case it seems to be an output-only so should better appear after |depfile_ins| for clarity.

moritzx22 replied:
see [comment](https://github.com/ninja-build/ninja/pull/2680#discussion_r3089414978)

digit-google commented on src/graph.h:
```diff
@@ -304,17 +305,18 @@ struct ImplicitDepLoader {
                                   std::vector<StringPiece>* depfile_ins,
                                   std::string* err);
 
   /// Load implicit dependencies for \a edge from a depfile attribute.
```
again, update documentation accordingly. Keep Edge* input and EdgeInputsRange* outputs separate for clarity if necessary.

moritzx22 replied:
see [comment](https://github.com/ninja-build/ninja/pull/2680#discussion_r3089414978)

digit-google commented on src/graph.h:
```diff
@@ -304,17 +305,18 @@ struct ImplicitDepLoader {
+  /// Load implicit dependencies for \a input_range from the DepsLog.
   /// @return false on error (without filling \a err if info is just missing).
-  bool LoadDepsFromLog(Edge* edge, std::string* err);
+  bool LoadDepsFromLog(EdgeInputsRange* input_range, std::string* err);
```
should this be a `const EdgeInputsRange*` or does the value mutates due to this function. Please clarify.

moritzx22 replied:
Parameter shifted to return value, see [comment](https://github.com/ninja-build/ninja/pull/2680#discussion_r3089414978).

digit-google commented on src/graph.h:
```diff
@@ -324,9 +326,30 @@ struct ImplicitDepLoader {
+  /// Create new instance covering all |edge| inputs.
+  EdgeInputsRange(Edge* edge)
+      : beg_(edge->inputs_.begin()), end_(edge->inputs_.end()),
+        edge_(edge) {}
```
Add an EdgeInputsRange(iterator, iterator) constructor and make beg_ and end_ private members.

moritzx22 replied:
Done.

moritzx22 commented:
### Previous push
rebased to master

### Next push
incorporated comments, see [diff](https://github.com/ninja-build/ninja/compare/9bfeee77518052b4f94f03cd690228ddfee94651..2d73512a1b7c7b319cfd2fa706b9387b69a7544a)

digit-google reviewed (approved):
This is good, the code makes so much more sense now. Thanks for your patience!
@jhasse please take a look.

moritzx22 commented:
Great to hear the code is clearer. Thanks for the reasonable comments!

moritzx22 commented:
next push is minor, see https://github.com/ninja-build/ninja/pull/2680#discussion_r3087518316

moritzx22 commented on src/graph.cc:
```diff
@@ -32,6 +32,211 @@
+  const bool isRestat_ = edge_->GetBindingBool("restat");
+  bool generator_ = false;
+  bool generatorValid_ = false;
+  std::vector<CachedLogEntry> logEntry_;
```
One additional memory alignment optimization is possible: the size of each `CachedLogEntry` element in `logEntry_` can be reduced from 16 to 8 bytes by replacing the bool `evaluated_` field with a sentinel pointer. This halves the vector's memory footprint and improves cache efficiency for edges with many outputs.

Whether this is worth merging is debatable, most edges have only 1–3 outputs, so the benefit might be limited in practice.

See the [branch](https://github.com/moritzx22/ninja/tree/fix2666_alignment) and the [diff](https://github.com/moritzx22/ninja/compare/fix2666..fix2666_alignment).

moritzx22 commented:
### Next Push
resolve conflicts, rebase to master.
